File: solutions/sheet-05/java/QGramIndexTest.java

// Copyright 2017, University of Freiburg,
// Chair of Algorithms and Data Structures.
// Author: Hannah Bast <bast@cs.uni-freiburg.de>,
// Claudius Korzen <korzen@cs.uni-freiburg.de>.

import static org.junit.Assert.assertEquals;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import org.junit.Test;

/**
 * One unit test for each non-trivial method in the QGramIndex class.
 */
public class QGramIndexTest {

  /**
   * Tests for the constructor.
   */
  @Test
  public void testConstructor() {
    QGramIndex qgi = new QGramIndex(3, false);
    assertEquals(3, qgi.q);
    assertEquals("$$", qgi.padding);
    assertEquals("{}", qgi.invertedLists.toString());
  }

  /**
   * Tests for the method buildFromFile().
   */
  @Test
  public void testBuildFromFile() {
    QGramIndex qgi = new QGramIndex(3, false);
    qgi.buildFromFile("example.tsv");
    assertEquals("{$$b=[2], $$f=[1], $br=[2], $fr=[1], bre=[2], fre=[1], "
        + "rei=[1, 2]}", qgi.invertedLists.toString());
    assertEquals("[Entity(name='frei', score=3, desc='a word'), "
        + "Entity(name='brei', score=2, desc='another word')]",
        qgi.entities.toString());
  }

  /**
   * Tests for the method mergeLists().
   */
  @Test
  public void testMergeLists() {
    List<Integer> first = Arrays.asList(1, 1, 3, 5);
    List<Integer> second = Arrays.asList(2, 3, 3, 9, 9);
    List<List<Integer>> lists = Arrays.asList(first, second);

    assertEquals("[(1, 2), (2, 1), (3, 3), (5, 1), (9, 2)]",
        QGramIndex.mergeLists(lists).toString());
  }

  /**
   * Tests for the method prefixEditDistance().
   */
  @Test
  public void testPrefixEditDistance() {
    assertEquals(0, QGramIndex.prefixEditDistance("frei", "frei", 0));
    assertEquals(0, QGramIndex.prefixEditDistance("frei", "freiburg", 0));
    assertEquals(1, QGramIndex.prefixEditDistance("frei", "breifurg", 1));
    assertEquals(3, QGramIndex.prefixEditDistance("freiburg", "stuttgart", 2));
  }

  /**
   * Tests for the method findMatches().
   */
  @Test
  public void testFindMatches() {
    QGramIndex qgi = new QGramIndex(3, false);
    qgi.buildFromFile("example.tsv");
    assertEquals("([Entity(name='frei', score=3, desc='a word', ped=0)], 1)",
        qgi.findMatches("frei", 0).toString());
    assertEquals("([Entity(name='frei', score=3, desc='a word', ped=0), "
        + "Entity(name='brei', score=2, desc='another word', ped=1)], 2)",
        qgi.findMatches("frei", 2).toString());
    assertEquals("([Entity(name='frei', score=3, desc='a word', ped=2)], 2)",
        qgi.findMatches("freibu", 2).toString());
  }

  /**
   * Tests for the method findMatches().
   */
  @Test
  public void testRankMatches() {
    List<Entity> matches = new ArrayList<>();

    Entity entity1 = new Entity("foo", 3);
    entity1.ped = 2;
    matches.add(entity1);

    Entity entity2 = new Entity("bar", 7);
    entity2.ped = 0;
    matches.add(entity2);

    Entity entity3 = new Entity("baz", 2);
    entity3.ped = 1;
    matches.add(entity3);

    Entity entity4 = new Entity("boo", 5);
    entity4.ped = 1;
    matches.add(entity4);

    assertEquals("[Entity(name='bar', score=7, ped=0), "
        + "Entity(name='boo', score=5, ped=1), "
        + "Entity(name='baz', score=2, ped=1), "
        + "Entity(name='foo', score=3, ped=2)]",
        QGramIndex.rankMatches(matches).toString());
  }

  /**
   * Tests for the method computeQGrams().
   */
  @Test
  public void testComputeQGrams() {
    QGramIndex qgi = new QGramIndex(3, false);
    assertEquals("[$$f, $fr, fre, rei, eib, ibu, bur, urg]",
        qgi.computeQGrams("freiburg").toString());
  }

  /**
   * Tests for the method normalize().
   */
  @Test
  public void testNormalize() {
    assertEquals("freiburg", QGramIndex.normalize("Frei, burg !!"));
    assertEquals("freiburg", QGramIndex.normalize("freiburg"));
  }
}

File: solutions/sheet-05/cpp/QGramIndexTest.cpp
// Copyright 2017, University of Freiburg,
// Chair of Algorithms and Data Structures.
// Authors: Björn Buchhold <buchholb@cs.uni-freiburg.de>,
//          Patrick Brosi <brosi@cs.uni-freiburg.de>,
//          Claudius Korzen <korzen@cs.uni-freiburg.de>.

#include <gtest/gtest.h>
#include "./QGramIndex.h"


// _____________________________________________________________________________
TEST(QGramIndexTest, testBuildFromFile) {
  QGramIndex index(3, false);
  index.buildFromFile("example.tsv");

  ASSERT_EQ(7, index._invertedLists.size());

  ASSERT_EQ(1, index._invertedLists["$$b"].size());
  ASSERT_EQ(2, index._invertedLists["$$b"][0]);
  ASSERT_EQ(1, index._invertedLists["$$f"].size());
  ASSERT_EQ(1, index._invertedLists["$$f"][0]);
  ASSERT_EQ(1, index._invertedLists["$br"].size());
  ASSERT_EQ(2, index._invertedLists["$br"][0]);
  ASSERT_EQ(1, index._invertedLists["$fr"].size());
  ASSERT_EQ(1, index._invertedLists["$fr"][0]);
  ASSERT_EQ(1, index._invertedLists["bre"].size());
  ASSERT_EQ(2, index._invertedLists["bre"][0]);
  ASSERT_EQ(1, index._invertedLists["fre"].size());
  ASSERT_EQ(1, index._invertedLists["fre"][0]);
  ASSERT_EQ(2, index._invertedLists["rei"].size());
  ASSERT_EQ(1, index._invertedLists["rei"][0]);
  ASSERT_EQ(2, index._invertedLists["rei"][1]);
}

// _____________________________________________________________________________
TEST(QGramIndexTest, testMergeLists) {
  std::vector<size_t> first { 1, 1, 3, 5 };
  std::vector<size_t> second { 2, 3, 3, 9, 9 };
  std::vector<const std::vector<size_t>* > lists { &first, &second };

  // std::vector<std::pair<size_t, size_t>> result = QGramIndex::mergeLists
  // (lists);

//  ASSERT_EQ(5, result.size());
//  ASSERT_EQ(1, result[0].first);
//  ASSERT_EQ(2, result[0].second);
//  ASSERT_EQ(2, result[1].first);
//  ASSERT_EQ(1, result[1].second);
//  ASSERT_EQ(3, result[2].first);
//  ASSERT_EQ(3, result[2].second);
//  ASSERT_EQ(5, result[3].first);
//  ASSERT_EQ(1, result[3].second);
//  ASSERT_EQ(9, result[4].first);
//  ASSERT_EQ(2, result[4].second);
}

// _____________________________________________________________________________
TEST(QGramIndexTest, testPrefixEditDistance) {
  ASSERT_EQ(0, QGramIndex::prefixEditDistance("frei", "frei", 0));
  ASSERT_EQ(0, QGramIndex::prefixEditDistance("frei", "freiburg", 0));
  ASSERT_EQ(1, QGramIndex::prefixEditDistance("frei", "breifurg", 1));
  ASSERT_EQ(3, QGramIndex::prefixEditDistance("freiburg", "stuttgart", 2));
}

// _____________________________________________________________________________
TEST(QGramIndexTest, testFindMatches) {
  QGramIndex index(3, false);
  index.buildFromFile("example.tsv");

  auto result0 = index.findMatches("frei", 0);
  ASSERT_EQ(1, result0.first.size());
  ASSERT_EQ("frei", result0.first[0].name);
  ASSERT_EQ(3, result0.first[0].score);
  ASSERT_EQ("a word", result0.first[0].description);
  ASSERT_EQ(0, result0.first[0].ped);
  ASSERT_EQ(1, result0.second);

  auto result1 = index.findMatches("frei", 2);
  ASSERT_EQ(2, result1.first.size());
  ASSERT_EQ("frei", result1.first[0].name);
  ASSERT_EQ(3, result1.first[0].score);
  ASSERT_EQ("a word", result1.first[0].description);
  ASSERT_EQ(0, result1.first[0].ped);
  ASSERT_EQ("brei", result1.first[1].name);
  ASSERT_EQ(2, result1.first[1].score);
  ASSERT_EQ("another word", result1.first[1].description);
  ASSERT_EQ(1, result1.first[1].ped);
  ASSERT_EQ(2, result1.second);

  auto result2 = index.findMatches("freibu", 2);
  ASSERT_EQ(1, result2.first.size());
  ASSERT_EQ("frei", result2.first[0].name);
  ASSERT_EQ(3, result2.first[0].score);
  ASSERT_EQ("a word", result2.first[0].description);
  ASSERT_EQ(2, result2.first[0].ped);
  ASSERT_EQ(2, result2.second);
}

// _____________________________________________________________________________
TEST(QGramIndexTest, testRankMatches) {
  std::vector<Entity> matches;

  Entity entity0("foo", 3);
  entity0.ped = 2;
  matches.push_back(entity0);

  Entity entity1("bar", 7);
  entity1.ped = 0;
  matches.push_back(entity1);

  Entity entity2("baz", 2);
  entity2.ped = 1;
  matches.push_back(entity2);

  Entity entity3("boo", 5);
  entity3.ped = 1;
  matches.push_back(entity3);

  std::vector<Entity> ranked = QGramIndex::rankMatches(matches);
  ASSERT_EQ(4, ranked.size());
  ASSERT_EQ("bar", ranked[0].name);
  ASSERT_EQ("boo", ranked[1].name);
  ASSERT_EQ("baz", ranked[2].name);
  ASSERT_EQ("foo", ranked[3].name);
}

// _____________________________________________________________________________
TEST(QGramIndexTest, testComputeQGrams) {
  QGramIndex index(3, false);
  auto qgrams = index.computeQGrams("freiburg");
  ASSERT_EQ(qgrams.size(), 8);
  ASSERT_EQ(qgrams[0], "$$f");
  ASSERT_EQ(qgrams[1], "$fr");
  ASSERT_EQ(qgrams[2], "fre");
  ASSERT_EQ(qgrams[3], "rei");
  ASSERT_EQ(qgrams[4], "eib");
  ASSERT_EQ(qgrams[5], "ibu");
  ASSERT_EQ(qgrams[6], "bur");
  ASSERT_EQ(qgrams[7], "urg");
}

// _____________________________________________________________________________
TEST(QGramIndexTest, testNormalizeString) {
  ASSERT_EQ("freiburg", QGramIndex::normalize("Frei, burg !!"));
  ASSERT_EQ("freiburg", QGramIndex::normalize("freiburg"));
}
